Add support for Setω+n

jespercockx opened the pull request:
This PR adds internal syntax and typing rules for a second universe hierarchy `Setω+n`, which sits on top of Agda's regular universe hierarchy. Unlike `Set i`, this hierarchy is not polymorphic, so it is not possible to quantify over all `Setω+n` at once. This is my proposed answer to #4595 and fixes issues #2119, #4109, and #4585.

Note that there is no concrete syntax yet for `Setω+n` (except in error messages). I tried to add it in the same way as I did for `Setω`, but adding an infinite number of builtins is hard. Instead, it might be necessary to add hardcoded support for it in the parser, unless someone can think of a better solution.

jespercockx committed:
[ #4595 ] Add support for Setω+n (no concrete syntax yet)

jespercockx committed:
[ fix #2119 ] Add test case

jespercockx committed:
[ fix #4109 ] Add test case

jespercockx committed:
[ fix #4585 ] Add test case

jespercockx committed:
[ sorts ] Add PTS rule SizeUniv : Setω

Together with the addition of Setω+i, this ensures that every closed Agda sort has a successor sort. Note that this change currently has no effect since SizeUniv is defined to be equal to Set0.

jespercockx committed:
[ re #4595 ] Remove (now superfluous) "Has bigger sort" constraint

dylanede commented:
I'm just trying this out, and fixing #4109 in particular has made it much easier to do some reasoning about reflected terms in macros, so thanks for that! In particular I can now have a data structure
```agda
data _Reflects_ : {a : Level} {A : Set a} → Term → A → Setω where
  ...
```
to express that a reflected term corresponds to a particular real term, which then makes it easier to safely manipulate and analyse real terms according to their structure.

dylanede commented:
Quick note:

When I try C-c C-c to case split on an instance of a data type in `Setω`, I get this error message:
```text
Cannot split on datatype in sort Setω
```
Is that something that should be changed?

**Edit:** Actually, it won't let me pattern match at all on such a datatype. It's not just the editor integration. I guess that needs to be fixed to make `data` types in `Setω` useful.

dylanede commented:
Would it just be a case of stopping it from triggering the error, in this code?
https://github.com/agda/agda/blob/1f17ae4b3e65d00645c1e012af423e1ab7ebcad8/src/full/Agda/TypeChecking/Rules/LHS.hs#L1759-L1773

jespercockx committed:
[ re #4595 ] Allow splitting on datatypes in Setω+n

jespercockx commented:
> Would it just be a case of stopping it from triggering the error, in this code?
> 
> https://github.com/agda/agda/blob/1f17ae4b3e65d00645c1e012af423e1ab7ebcad8/src/full/Agda/TypeChecking/Rules/LHS.hs#L1759-L1773

Indeed, that's it. I added a fix.

jespercockx commented:
So I noticed that the printing of `Setω+n` is currently quite horribly broken. In particular, if you have an identifier in scope that's named `Setω+1`, then Agda will print the *sort* `Setω+1` as `Setω+2`. The reason is that the printing currently uses a hack to print these sorts by printing them as a variable, but if another identifier with the same name is in scope then the shadowing machinery kicks in to rename the variable. Actually the same problem already appears for other built-in sorts `piSort`, `funSort`, and `univSort`. Maybe we should finally add a mechanism for binding built-in sorts so we can add all of these as proper builtins to `Agda.Primitive`.

jespercockx commented:
I wonder if there would be some way to make `Set_n` and `Prop_n` into builtins as well, since this would allow renaming them when importing `Agda.Primitive`. Then perhaps every Agda file should start with an implicit `open import Agda.Primitive using (Set; Prop)` to preserve the current behaviour. The tricky part is to deal with the suffixes in `Set0`, `Set1`, `Set2`, etc. We would first need to implement a mechanism for parsing name variants.

nad commented:
See #4093.

andreasabel commented on src/full/Agda/Syntax/Translation/InternalToAbstract.hs:
```diff
@@ -1305,9 +1305,12 @@ instance Reify Sort Expr where
           I.Def inf [] <- fromMaybe __IMPOSSIBLE__ <$> getBuiltin' builtinSetOmega
           return $ A.Def inf
+        I.Inf n -> do
+          infs <- freshName_ ("Setω+" ++ show n :: String) -- TODO: hack
```
Generalize the `A.Set` constructor of `A.Expr` to accommodate the new levels.

andreasabel commented on src/full/Agda/TypeChecking/Conversion.hs:
```diff
@@ -609,9 +609,9 @@ compareAtom cmp t m n =
-              equalType (El Inf $ apply tSub $ a : map (setHiding NotHidden) [bA,phi,u])
-                        (El Inf $ apply tSub $ a : map (setHiding NotHidden) [bA',phi',u'])
-              compareAtom cmp (AsTermsOf $ El Inf $ apply tSub $ a : map (setHiding NotHidden) [bA,phi,u])
+              equalType (El (Inf 0) $ apply tSub $ a : map (setHiding NotHidden) [bA,phi,u])
```
Looks hackish.

jespercockx replied:
I agree. I'm not sure why `Inf` is needed here in the first place, but I guess it is because there is no dedicated sort for non-fibrant types like the interval / face lattice of cubical. Maybe @Saizan can comment on this.

Saizan replied:
`Inf` is the sort of `tSub` and a few other cubical primitives.

Indeed that's because it's outside the range of sorts indexed by a `Level`, which are the ones with composition, is that still the case after this PR?

jespercockx replied:
It's still the case that `Inf` is outside of the range of sorts that are indexed by `Level`, but it is no longer disallowed to split on types in it. So it would probably be good to move the interval to its own sort if this PR is merged.

andreasabel commented on src/full/Agda/TypeChecking/Conversion.hs:
```diff
@@ -609,9 +609,9 @@ compareAtom cmp t m n =
-              compareAtom cmp (AsTermsOf $ El Inf $ apply tSub $ a : map (setHiding NotHidden) [bA,phi,u])
+              equalType (El (Inf 0) $ apply tSub $ a : map (setHiding NotHidden) [bA,phi,u])
+                        (El (Inf 0) $ apply tSub $ a : map (setHiding NotHidden) [bA',phi',u'])
+              compareAtom cmp (AsTermsOf $ El (Inf 0) $ apply tSub $ a : map (setHiding NotHidden) [bA,phi,u])
```
Could there be a special constructor replacing the former `AsTermsOf . El Inf` ?

jespercockx replied:
Yes, I'll add a pattern synonym.

jespercockx replied:
On second thought, doesn't really seem worth it adding a pattern synonym for a single use.

andreasabel commented on src/full/Agda/TypeChecking/Conversion.hs:
```diff
@@ -1229,13 +1236,12 @@ leqSort s1 s2 = (catchConstraint (SortCmp CmpLeq s1 s2) :: m () -> m ()) $ do
-      (UnivSort Inf , UnivSort Inf) -> yes
+      -- If the first sort is a small sort that rigidly depends on a
+      -- variable and the second sort does not mention this variable,
+      -- the second sort must be at least Inf.
```
... at least Setomega

andreasabel commented on src/full/Agda/TypeChecking/Errors.hs:
```diff
@@ -536,10 +536,10 @@ instance PrettyTCM TypeError where
-      (s            , Sort MetaS{} ) -> prettyTCM $ ShouldBeASort $ El Inf s
-      (Sort DefS{}  , t            ) -> prettyTCM $ ShouldBeASort $ El Inf t
-      (s            , Sort DefS{}  ) -> prettyTCM $ ShouldBeASort $ El Inf s
+      (Sort MetaS{} , t            ) -> prettyTCM $ ShouldBeASort $ El (Inf 0) t
```
Consider `__IMPOSSIBLE__` instead of 0.

andreasabel commented on src/full/Agda/TypeChecking/Rules/LHS.hs:
```diff
@@ -1760,14 +1760,13 @@ checkSortOfSplitVar :: (MonadTCM m, MonadReduce m, MonadError TCErr m, ReadTCSta
       | Just target <- mtarget -> unlessM (isPropM target) splitOnPropError
       | otherwise              -> splitOnPropError
-    Inf{} | infOk -> return ()
+    Inf{} -> return ()
```
Is there an issue you could link to?

andreasabel commented on src/full/Agda/TypeChecking/Serialise.hs:
```diff
@@ -72,7 +72,7 @@ import Agda.Utils.Except
 
 currentInterfaceVersion :: Word64
-currentInterfaceVersion = 20200411 * 10 + 0
+currentInterfaceVersion = 20200418 * 10 + 0
```
This will be outdated at the time of merging.

andreasabel commented on src/full/Agda/TypeChecking/Substitute.hs:
```diff
@@ -1427,34 +1427,38 @@ instance (Subst t a, Ord a) => Ord (Elim' a) where
-  ifM ((optOmegaInOmega <$> pragmaOptions) `or2M` typeInType)
-  {-then-} (return $ Just Inf)
-  {-else-} (return Nothing)
+univSort' :: Sort -> Maybe Sort
```
Consider renaming to `succSort'`.  Reading through the PR I always wondered what `univSort` meant.  

But then, the resp. constructor should be renamed as well.

pnlph replied:
Or maybe `sucSort'`? `suc`seems to be the standard notation for the successor in Agda.

jespercockx replied:
This doesn't really have anything to do with this PR, feel free to file a separate issue if you want to rename `univSort` to something else. I actually prefer it over `succSort` since it is more descriptive: it is the sort of a universe.

andreasabel commented on src/full/Agda/TypeChecking/Substitute.hs:
```diff
@@ -1427,34 +1427,38 @@ instance (Subst t a, Ord a) => Ord (Elim' a) where
+isSmallSort Type{}   = True
+isSmallSort Prop{}   = True
+isSmallSort SizeUniv = True
+isSmallSort _        = False
```
Expand catch-all.

andreasabel reviewed (changes requested):
LGTM.
I do not grasp the change in its entirety, but we'll see it once people start using it.

For printing, we should consider using subscripts like for `Set_1`: `Set\omega_1` etc.

pnlph commented:
> I wonder if there would be some way to make `Set_n` and `Prop_n` into builtins as well, since this would allow renaming them when importing `Agda.Primitive`. Then perhaps every Agda file should start with an implicit `open import Agda.Primitive using (Set; Prop)` to preserve the current behaviour. The tricky part is to deal with the suffixes in `Set0`, `Set1`, `Set2`, etc. We would first need to implement a mechanism for parsing name variants.

Now `Prop` only works when the option `--prop` is enabled, right? (I was checking with 2.6.1), so maybe it is only necessary to use `open import Agda.Primitive using (Set)` if the current functionality is to be preserved.

But how will the user rename `Set`if it is already implicitly imported?

An alternative would be to change the current behaviour of the empty module and allowing the old behaviour with a new [import command-line option](https://agda.readthedocs.io/en/latest/tools/command-line-options.html?highlight=options#imports-and-libraries) maybe `-- classic` or something like that ;-) to activate the implicit import.

jespercockx committed:
[ refactor ] Add third 'unknown' alternative to isSmallSort

jespercockx committed:
[ refactor ] __IMPOSSIBLE__ instead of Inf 0

jespercockx committed:
[ re #4109 ] Add reference to issue in comment

Saizan commented:
> > Would it just be a case of stopping it from triggering the error, in this code?
> > https://github.com/agda/agda/blob/1f17ae4b3e65d00645c1e012af423e1ab7ebcad8/src/full/Agda/TypeChecking/Rules/LHS.hs#L1759-L1773
> 
> Indeed, that's it. I added a fix.

Can you make sure it's still impossible to split on the `Interval`?

jespercockx committed:
[ #4595 ] Add concrete syntax + proper pretty printing for Setωᵢ

jespercockx commented:
> > 
> 
> Can you make sure it's still impossible to split on the `Interval`?

The most robust way to do this would be to give the interval its own sort.

jespercockx commented:
I rebased this PR on the current master after the fix of #4093, and added the concrete syntax `Setωᵢ` as suggested by Andreas. The main thing still blocking this from being merged is that it possibly makes Cubical Agda inconsistent (?), since it enables splitting on types in `Setω`, so in particular on the interval. The proper solution would be to have a separate universe of non-fibrant types (see https://github.com/agda/agda/pull/4091). An alternative would be to add a manual check to the LHS checker that prevents matching on the constructors of the interval directly.

Saizan commented:
Actually the coverage checker is forbidding splitting on the interval in the `isDatatype` function, this was necessary to avoid problems with `--omega-in-omega` already.

jespercockx commented:
Ok, I was already suspecting something like that because there were no failing tests. So I will go ahead and merge this then.

jespercockx committed:
[ re #4595 ] Update some comments and debug output

jespercockx committed:
[ #2119 #4585 ] Add entry to changelog

gallais commented:
I am hyped. Thanks @jespercockx for this PR!
